server: add ./cockroach mt start-sql

tbg opened the pull request:
This adds a CLI command to start a SQL tenant in a standalone process.

The tenant currently communicates with the KV layer "as a node"; this
will only change with #47898. As is, the tenant has full access to the
KV layer and so a few things may work that shouldn't as a result of
that.

Additionally, the tenant runs a Gossip client as a temporary stopgap
until we have removed the remaining dependencies on it (#49693 has
the details on what those are).

Apart from that, though, this is the real thing and can be used to
start setting up end-to-end testing of ORMs as well as the deploy-
ments.

A word on the choice of the `mt` command: `mt` is an abbreviation for
`multi-tenant` which was considered too long; just `tenant` was
considered misleading - there will be multiple additional subcommands
housing the other services required for running the whole infrastructure
including certs generation, the KV auth broker server, and the SQL
proxy. Should a lively bikeshed around naming break out we can rename
the commands later when consensus has been reached.

For those curious to try it out the following will be handy:

```bash
rm -rf ~/.cockroach-certs cockroach-data &&
killall -9 cockroach && \
export COCKROACH_CA_KEY=$HOME/.cockroach-certs/ca.key && \
./cockroach cert create-ca && \
./cockroach cert create-client root && \
./cockroach cert create-node 127.0.0.1 && \
./cockroach start --host 127.0.0.1 --background && \
./cockroach sql --host 127.0.0.1 -e 'select crdb_internal.create_tenant(123);'
./cockroach mt start-sql --tenant-id 123 --kv-addrs 127.0.0.1:26257 --sql-addr :5432 &
sleep 1 && \
./cockroach sql --host 127.0.0.1 --port 5432
```

This roughly matches what the newly introduced `acceptance/multitenant`
roachtest does as well.

cc @nvanbenschoten @asubiotto 

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49908)
<!-- Reviewable:end -->

knz reviewed (commented):
> Reviewed 8 of 8 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49908)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/flags.go, line 726 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93BZ359peWH6usiAVy:-M93BZ359peWH6usiAVz:bgaowy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cli/flags.go#L726)):*
> ```Go
> 		// NB: serverInsecure populates baseCfg.{Insecure,SSLCertsDir} in this the following method
> 		// (which is a PreRun for this command):
> 		_ = extraServerFlagInit
> ```

I think that's missing `...()` at the end
___
*[pkg/cli/flags.go, line 732 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93Bb5n82Po4te6um6-:-M93Bb5n82Po4te6um60:b-qzxonq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cli/flags.go#L732)):*
> ```Go
> 		VarFlag(f, addrSetter{&serverSQLAddr, &serverSQLPort}, cliflags.ListenSQLAddr)
> 
> 		StringSlice(f, &serverCfg.SQLConfig.TenantKVAddrs, cliflags.KVAddrs, []string{"127.0.0.1:26257"})
> ```

the default value should be assigned to TenantKVAddrs in context.go
___
*[pkg/cmd/roachtest/multitenant.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93Bz1F6yibQGZ9sT3u:-M93Bz1F6yibQGZ9sT3v:b-nplyo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cmd/roachtest/multitenant.go#L41)):*
> ```Go
> 			"--tenant-id", "123",
> 			"--kv-addrs", strings.Join(kvAddrs, ","),
> 			"--sql-addr", "0.0.0.0:36257",
> ```

I don't understand "0.0.0.0" here - why do you need to listen on all interfaces?


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49908)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cli/flags.go, line 726 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93BZ359peWH6usiAVy:-M94Ba3e6AxjuTCDZwJK:bx5z5ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cli/flags.go#L726)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think that's missing `...()` at the end
</blockquote></details>

No, I didn't want to call it - this is just an anchor to make it easier to navigate to the method (and it prevents rot should the method get renamed). 
___
*[pkg/cli/flags.go, line 732 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93Bb5n82Po4te6um6-:-M94DolkDQuA8ImeIYem:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cli/flags.go#L732)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

the default value should be assigned to TenantKVAddrs in context.go
</blockquote></details>

Done.
___
*[pkg/cmd/roachtest/multitenant.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93Bz1F6yibQGZ9sT3u:-M94Dphq8kmi4m2MADpJ:bay3xqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cmd/roachtest/multitenant.go#L41)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't understand "0.0.0.0" here - why do you need to listen on all interfaces?
</blockquote></details>

I don't, but I want to make sure I listen on the external one in case this is running in roachtest. What do you want to see here? I can make it `127.0.0.1` if local if that's the concern. (Done).


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
I would recommend ensuring that the output of `cockroach mt help` and the `--help` outputs for the sub-commands highlight that the feature is experimental (our code word for "we don't support this externally")

LGTM otherwise

> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-:-M94amhnD8pyfNs7ilOU:b99crth)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/cmd/roachtest/multitenant.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93Bz1F6yibQGZ9sT3u:-M94aWt725t4DsfRJthw:bf0zl0u) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cmd/roachtest/multitenant.go#L41)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

I don't, but I want to make sure I listen on the external one in case this is running in roachtest. What do you want to see here? I can make it `127.0.0.1` if local if that's the concern. (Done).
</blockquote></details>

Ok I simply did not understand.
In that case, I'd suggest simplifying to `"--sql-addr", ":36257"` without an address prefix, in which case it should default to all-interfaces already.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> I would recommend ensuring that the output of cockroach mt help and the --help outputs for the sub-commands highlight that the feature is experimental (our code word for "we don't support this externally")

Done (the `mt` command is also hidden).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-:-M95ARzO0KqnGi5xDEy6:b-nf92pe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/cmd/roachtest/multitenant.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M93Bz1F6yibQGZ9sT3u:-M957GO5BYMtDlYhRYL9:b-3boif8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0dbe16f11f755539ebd8bf2a689376491e12cb3d/pkg/cmd/roachtest/multitenant.go#L41)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ok I simply did not understand.
In that case, I'd suggest simplifying to `"--sql-addr", ":36257"` without an address prefix, in which case it should default to all-interfaces already.
</blockquote></details>

Are you opposed to the current code? I think it's kind of nice to not listen everywhere on `--local` runs.


<!-- Sent from Reviewable.io -->

knz commented:
> Are you opposed to the current code? I think it's kind of nice to not listen everywhere on --local runs.

No objection.

nvb reviewed (approved):
Nothing major in the review. This is exciting to see!

> Reviewed 3 of 8 files at r1, 4 of 4 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-:-M95hhEl7UP-u093dWSw:b6tqlyb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/cli_mt.go, line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95eH8q3W8PCcGWYrgu:-M95eH8r7Jr3PFfrVT5t:b-xcerkp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/cli/cli_mt.go#L1)):*
> ```Go
> // Copyright 2020 The Cockroach Authors.
> ```

nit: why prefix the filename with `cli_`? We don't seem to do that anywhere else.
___
*[pkg/cli/cli_mt.go, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95eijwBcKrKM63psOh:-M95eijwBcKrKM63psOi:brrzvlp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/cli/cli_mt.go#L16)):*
> ```Go
> 
> func init() {
> 	cockroachCmd.AddCommand(mtCmd)
> ```

I'm assuming you deliberately did not place this in the list of all other commands in `cli.go`?
___
*[pkg/cli/flags.go, line 741 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95fDNf4VTdlfmgnI76:-M95fDNf4VTdlfmgnI77:bgl05w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/cli/flags.go#L741)):*
> ```Go
> 		return errors.Wrap(err, "invalid tenant ID")
> 	}
> 	*w.tenID = roachpb.MakeTenantID(tenID)
> ```

This is going to blow up with a tenID of 0. Are you ok with that?
___
*[pkg/cli/cliflags/flags_mt.go, line 23 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95faYw89u3Seir6NTV:-M95faYw89u3Seir6NTW:b-r1b8sw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/cli/cliflags/flags_mt.go#L23)):*
> ```Go
> 	KVAddrs = FlagInfo{
> 		Name:        "kv-addrs",
> 		Shorthand:   "",
> ```

nit: why specify this?
___
*[pkg/cmd/roachtest/acceptance.go, line 45 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95fmIb0eJX_11wlUU-:-M95fmIb0eJX_11wlUU0:bzifhg1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/cmd/roachtest/acceptance.go#L45)):*
> ```Go
> 		{name: "gossip/restart-node-one", fn: runGossipRestartNodeOne},
> 		{name: "gossip/locality-address", fn: runCheckLocalityIPAddress},
> 		{name: "multitenant", fn: runAcceptanceMultitenant},
> ```

We need a `minVersion` here, right?
___
*[pkg/server/testserver.go, line 629 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95gTK711GwAbDck556:-M95gTK711GwAbDck557:b-bqn2gd) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/server/testserver.go#L629)):*
> ```Go
> }
> 
> // StartTenant starts a stand-alone SQL server against a dedicated KV backend.
> ```

What does "dedicated" mean here? The KV backend can and is still running its own SQL server, right?


<!-- Sent from Reviewable.io -->

tbg committed:
server: add ./cockroach mt start-sql

This adds a CLI command to start a SQL tenant in a standalone process.

The tenant currently communicates with the KV layer "as a node"; this
will only change with #47898. As is, the tenant has full access to the
KV layer and so a few things may work that shouldn't as a result of
that.

Additionally, the tenant runs a Gossip client as a temporary stopgap
until we have removed the remaining dependencies on it (#49693 has
the details on what those are).

Apart from that, though, this is the real thing and can be used to
start setting up end-to-end testing of ORMs as well as the deploy-
ments.

A word on the choice of the `mt` command: `mt` is an abbreviation for
`multi-tenant` which was considered too long; just `tenant` was
considered misleading - there will be multiple additional subcommands
housing the other services required for running the whole infrastructure
including certs generation, the KV auth broker server, and the SQL
proxy. Should a lively bikeshed around naming break out we can rename
the commands later when consensus has been reached.

For those curious to try it out the following will be handy:

```bash
rm -rf ~/.cockroach-certs cockroach-data &&
killall -9 cockroach && \
export COCKROACH_CA_KEY=$HOME/.cockroach-certs/ca.key && \
./cockroach cert create-ca && \
./cockroach cert create-client root && \
./cockroach cert create-node 127.0.0.1 && \
./cockroach start --host 127.0.0.1 --background && \
./cockroach sql --host 127.0.0.1 -e 'select crdb_internal.create_tenant(123);'
./cockroach mt start-sql --tenant-id 123 --kv-addrs 127.0.0.1:26257 --sql-addr :5432 &
sleep 1 && \
./cockroach sql --host 127.0.0.1 --port 5432
```

This roughly matches what the newly introduced `acceptance/multitenant`
roachtest does as well.

Release note: None

tbg reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

TFTR! Addressed all of the comments.

bors r=knz,nvanbenschoten

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-:-M9SRns50nUo0TN47nTA:bj7gmac)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @knz and @nvanbenschoten)

___
*[pkg/cli/flags.go, line 741 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95fDNf4VTdlfmgnI76:-M9SR8OL61PX692NdYdz:b-40k3ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/cli/flags.go#L741)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is going to blow up with a tenID of 0. Are you ok with that?
</blockquote></details>

I'm basically OK with it but I added a check to make things look nice.
___
*[pkg/cli/cli_mt.go, line 16 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49908#-M95eijwBcKrKM63psOh:-M9SQjWE8ztR-DWyL1zK:btqqu5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/5af4621c54ad85fa414e582fad63a670379c848b/pkg/cli/cli_mt.go#L16)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm assuming you deliberately did not place this in the list of all other commands in `cli.go`?
</blockquote></details>

Correct.


<!-- Sent from Reviewable.io -->
